src/cache.ts

```diff
@@ -0,0 +1,110 @@
+const CACHE_MATCHED_KEY = 'cache-matched-key';
+const CACHE_KEY_PREFIX = 'setup-java';
+
+interface PackageManager {
+  id: 'maven' | 'gradle';
+  /**
+   * Paths of the file that specify the files to cache.
+   */
+  path: string[];
+  pattern: string[];
+}
+const supportedPackageManager: PackageManager[] = [
+  {
+    id: 'maven',
+    path: [join(os.homedir(), '.m2', 'repository')],
+    // https://github.com/actions/cache/blob/0638051e9af2c23d10bb70fa9beffcad6cff9ce3/examples.md#java---maven
+    pattern: ['**/pom.xml']
+  },
+  {
+    id: 'gradle',
+    path: [join(os.homedir(), '.gradle', 'caches'), join(os.homedir(), '.gradle', 'wrapper')],
+    // https://github.com/actions/cache/blob/0638051e9af2c23d10bb70fa9beffcad6cff9ce3/examples.md#java---gradle
+    pattern: ['**/*.gradle*', '**/gradle-wrapper.properties']
+  }
+];
+
+function findPackageManager(id: string): PackageManager {
+  const packageManager = supportedPackageManager.find(packageManager => packageManager.id === id);
+  if (packageManager === undefined) {
+    throw new Error(`unknown package manager specified: ${id}`);
+  }
+  return packageManager;
+}
+
+/**
+ * A function that generates a cache key to use.
+ * Format of the generated key will be "${{ platform }}-${{ id }}-${{ fileHash }}"".
+ * If there is no file matched to {@link PackageManager.path}, the generated key ends with a dash (-).
+ * @see {@link https://docs.github.com/en/actions/guides/caching-dependencies-to-speed-up-workflows#matching-a-cache-key|spec of cache key}
+ */
+async function computeCacheKey(packageManager: PackageManager) {
+  const hash = await glob.hashFiles(packageManager.pattern.join('\n'));
+  return `${CACHE_KEY_PREFIX}-${process.env['RUNNER_OS']}-${packageManager.id}-${hash}`;
+}
+
+/**
+ * Restore the dependency cache
+ * @param id ID of the package manager, should be "maven" or "gradle"
+ */
+export async function restore(id: string) {
+  const packageManager = findPackageManager(id);
+  const primaryKey = await computeCacheKey(packageManager);
+
+  core.debug(`primary key is ${primaryKey}`);
+  core.saveState(STATE_CACHE_PRIMARY_KEY, primaryKey);
+  if (primaryKey.endsWith('-')) {
+    core.warning(
+      `No file in ${process.cwd()} matched to [${
+        packageManager.pattern
+      }], make sure you have checked out the target repository`
+    );
+    return;
+  }
+
+  const matchedKey = await cache.restoreCache(packageManager.path, primaryKey, [
+    `${CACHE_KEY_PREFIX}-${process.env['RUNNER_OS']}-${id}`
+  ]);
+  if (matchedKey) {
+    core.saveState(CACHE_MATCHED_KEY, matchedKey);
+    core.info(`Cache restored from key: ${matchedKey}`);
+  } else {
+    core.info(`${packageManager.id} cache is not found`);
+  }
+}
+
+/**
+ * Save the dependency cache
+ * @param id ID of the package manager, should be "maven" or "gradle"
+ */
+export async function save(id: string) {
+  const packageManager = findPackageManager(id);
+  const primaryKey = await computeCacheKey(packageManager);
```

Comment:
Hello @KengoTODA. Could you please change `const primaryKey = await computeCacheKey(packageManager);` to `const primaryKey = core.getState(STATE_CACHE_PRIMARY_KEY);`

Reply:
OK, I've confirmed that the [`actions/cache` also use the state](https://github.com/actions/cache/blob/5807af2642b6ffc80df306359122fd0ff9b571b8/src/save.ts#L30-L31), and seems nice then I'll change it so. :)
Thank you!

Reply:
done by https://github.com/actions/setup-java/pull/193/commits/7fe6c4d63194dfa8e9a1e3c8bdcd5f08b63ad5f9 👍